Add tf modules

RoboDoig opened the pull request:
Functionality for remaining Tinkerforge modules. Not yet physically tested on some of the modules but should work.

RoboDoig committed:
Air quality module

RoboDoig committed:
Air quality module

RoboDoig committed:
Particulate matter bricklet

RoboDoig committed:
Particulate matter bricklet

RoboDoig committed:
Sound pressure bricklet

RoboDoig committed:
Sound pressure bricklet

RoboDoig committed:
Industrial PTC bricklet

RoboDoig committed:
Industrial PTC bricklet

RoboDoig committed:
Cleanup - more explicit inheritance

RoboDoig committed:
Cleanup - more explicit inheritance

RoboDoig committed:
Humidity bricklet

RoboDoig committed:
Humidity bricklet

RoboDoig committed:
Prototype GPS module (coordinate)
Might want to squeeze more data out of this rather than having multiple GPS nodes

RoboDoig committed:
Prototype GPS module (coordinate)
Might want to squeeze more data out of this rather than having multiple GPS nodes

RoboDoig committed:
Prototype industrial analog out bricklet

RoboDoig committed:
Prototype industrial analog out bricklet

RoboDoig committed:
Cleanup

RoboDoig committed:
Cleanup

RoboDoig committed:
Enum naming for XML serialization

RoboDoig committed:
Enum naming for XML serialization

RoboDoig committed:
gps comments

RoboDoig committed:
gps comments

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+        [Description("Specifies the behavior of the Out Status LED.")]
+        public OutLedStatusConfig OutLedStatus { get; set; } = OutLedStatusConfig.Intensity;
+        public int OutLedStatusMin { get; set; } = 0;
+        public int OutLedStatusMax { get; set; } = 24000;
```
Add blank lines to space each property.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+                });
+
+                // Set voltage when input is in range
+                //return signal.Where(value => value > 0 && value < 10000).Do(value => { device.SetVoltage(value); });
```
I would expect their drivers to throw an exception or otherwise deal with this case correctly. I can confirm whether this is the case, but until then I would be happy to just remove this range filter.

glopesdev commented on Bonsai.Tinkerforge/IndustrialPTC.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Bonsai.Tinkerforge
+{
+    internal class IndustrialPTC
```
This looks like a placeholder class? If not necessary, we should probably just remove it.

glopesdev commented on Bonsai.Tinkerforge/BrickletSoundPressure.cs:
```diff
@@ -0,0 +1,84 @@
+        public long Period { get; set; } = 1000;
+
+        [Description("Specifies the size, resolution and bin size of the FFT.")]
+        public FftSize FFTSize { get; set; } = FftSize.FftSize512;
```
Rename property name to `FftSize`.

glopesdev commented on Bonsai.Tinkerforge/BrickletSoundPressure.cs:
```diff
@@ -0,0 +1,84 @@
+namespace Bonsai.Tinkerforge
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures sound pressure and specture from a Sound Pressure Level Bricklet")]
```
Slight rewording for clarity (include the period at the end of the sentence): Measures sound pressure level, in decibels, from a Sound Pressure Level Bricklet.

glopesdev commented on Bonsai.Tinkerforge/BrickletSoundPressure.cs:
```diff
@@ -0,0 +1,84 @@
+                {
+                    device.SetStatusLEDConfig((byte)StatusLed);
+                    device.SetConfiguration((byte)FFTSize, (byte)Weighting);
+                    device.SetDecibelCallbackConfiguration(Period, false, 'x', 0, 0);
```
Not super relevant, as the value is ignored, but for consistency with other bricklet implementations, might be nicer to set min = 0 and max = 1.

glopesdev commented on Bonsai.Tinkerforge/BrickletSoundPressure.cs:
```diff
@@ -0,0 +1,84 @@
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures sound pressure and specture from a Sound Pressure Level Bricklet")]
+    public class BrickletSoundPressure : Combinator<IPConnection, int>
```
Rename to `BrickletSoundPressureLevel` to match Tinkerforge name.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialPTC.cs:
```diff
@@ -0,0 +1,79 @@
+namespace Bonsai.Tinkerforge
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures temperature with Pt100 and Pt1000 sensors from a PTC Industrial Bricklet")]
```
Include period at end of sentence.

glopesdev replied:
Still seems to be missing, maybe this was lost in the rebase?

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialPTC.cs:
```diff
@@ -0,0 +1,79 @@
+        public int MovingAverageLengthTemperature { get; set; } = 1;
+
+        [Range(1, 1000)]
+        [Description("Specifies the moving average window length for resitance. A value of 1 turns off averaging.")]
```
Fix typo: resistance

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialPTC.cs:
```diff
@@ -0,0 +1,79 @@
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures temperature with Pt100 and Pt1000 sensors from a PTC Industrial Bricklet")]
+    public class BrickletIndustrialPTC : Combinator<IPConnection, int>
```
This is just a clarification comment. In this case we will opt to leave the acronym PTC capitalized, since it matches the underlying Tinkerforge API, so it is fine as it is.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+﻿using Bonsai.Expressions;
+using System;
+using System.ComponentModel;
+using System.Reactive.Disposables;
```
Remove unused namespaces.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,85 @@
+        [Description("Specifies the behavior of the status LED.")]
+        public BrickletGPSV2StatusLedConfig StatusLed { get; set; } = BrickletGPSV2StatusLedConfig.ShowStatus;
+
+        // TODO - can we get all the GPS data structures output here? GetStatus, GetAltitude, GetMotion, GetDateTime, GetSatelliteSystemStatus - could generate several observables for each Get function and then Merge
```
This is a very good point. I guess there are several options here, but to simplify we could check if it is possible to configure the period callbacks for all streams to the same value, and if they are reliably reported from the device. If this is the case, I guess we could `Zip` the outputs of all streams together and produce a combined DataFrame with all the data.

Another option is to have further operators downstream that select the individual callbacks, much as we do for `CreateIPConnection` and the individual bricklets.

RoboDoig replied:
While testing this I realised that having multiple nodes of the same type with the same Uid will crash Bonsai with the Device replaced exception. This happens in the initialisation when the configs are set.

Do we want to be able to have multiple nodes with the same Uid? This would be useful for GPS, since we can have one for coordinate, one for altitude etc. all pointing to the same GPS bricklet Uid. However, this will get tricky quite quickly when designing workflows. E.g. imagine having two ambient light nodes with the same bricklet Uid but with mistakenly different config parameters - which parameters will be applied?

RoboDoig replied:
See [c233a72]

glopesdev replied:
Yes, I can see how that would happen. I think we would rather have just one GPS operator that you can then store in a subject somewhere and then pull the different streams on demand. Let's discuss a bit about this offline and we can converge on a solution.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,85 @@
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using Tinkerforge;
+
```
Remove blank line.

glopesdev commented on Bonsai.Tinkerforge/BrickletAirQuality.cs:
```diff
@@ -0,0 +1,90 @@
+namespace Bonsai.Tinkerforge
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measure IAQ (indoor air quality) index, air pressure, humidity and temperature from an Air Quality Bricklet")]
```
Include missing period at end of sentence.

glopesdev commented on Bonsai.Tinkerforge/BrickletAirQuality.cs:
```diff
@@ -0,0 +1,90 @@
+        [Description("Specifies a temperature offset, in hundredths of a degree, to compensate for heat inside an enclosure.")]
+        public int TemperatureOffset { get; set; }
+
+        [Description("Specifies the duration of background calibration (requires power cycle to change)")]
```
Include missing period at end of sentence.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace Bonsai.Tinkerforge
+{
+    public class BrickletGPSV2 : Combinator<IPConnection, BrickletGPSV2.CoordinateData>
```
Add class description.

glopesdev commented on Bonsai.Tinkerforge/BrickletParticulateMatter.cs:
```diff
@@ -0,0 +1,73 @@
+namespace Bonsai.Tinkerforge
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures different sizes of particulate matter from a Particulate Matter Bricklet")]
```
Include missing period at end of sentence.

RoboDoig committed:
Requested changes for PR

RoboDoig committed:
Requested changes for PR

RoboDoig committed:
Proposed solution for GPS module

RoboDoig committed:
Proposed solution for GPS module

RoboDoig committed:
Test timer

RoboDoig committed:
Implementation comments

RoboDoig committed:
Add support for automatic search of bricklet UIDs (#2)

* Prototype automatic search for UIDs
* Better encapsulation of device data
* DeviceData helper
* Use Dictionary for TypeConverter
* Add auto-id to CO2 module
* Refactor type conversion for Bricklet Uid

Co-authored-by: Andrew Erskine <a.erskine@neurogears.org>

RoboDoig committed:
Final implementation of GPS Module
Some considerations: in order to get all the data out of the GPS module we generate several streams based on each of the data callbacks available. We then merge with a CombineLatest. In some cases (e.g. where no satellite fix) not all of the streams will return a value, in which case we will get no data at all from the merged stream. The workaround is to initialize all streams with an empty value to ensure that if only one callback is responding we always get some data out.
The creation of the streams at the moment uses alot of repetitive code. It's hard to avoid this since there are various event handlers for each data stream, and no unified way of setting callbackperiod on each device stream. There is a potential solution in gps-struct-test branch that attempts to make this more generic, but it is arguable much less readable than the current solution.

RoboDoig committed:
Extending device name lookup dict
And make sure all devices have typeconverter

RoboDoig committed:
Remove debug line

RoboDoig committed:
Resolve remote merge

glopesdev commented on Bonsai.Tinkerforge/BrickletAirQuality.cs:
```diff
@@ -0,0 +1,91 @@
+    public class BrickletAirQuality : Combinator<IPConnection, BrickletAirQuality.DataFrame>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Can we move the `TypeConverter` line to above the description? This is entirely aesthetical, but since `Description` attributes tend to always be longer strings, they sit more comfortably close to the property name. It might also help associating the description with the property when reading.

glopesdev commented on Bonsai.Tinkerforge/BrickletAmbientLightV3.cs:
```diff
@@ -10,7 +10,8 @@ namespace Bonsai.Tinkerforge
     {
-        [Description("The unique bricklet device UID.")]
+        [Description("Device data including address UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletAnalogInV3.cs:
```diff
@@ -0,0 +1,59 @@
+    public class BrickletAnalogInV3 : Combinator<IPConnection, int>
+    {
+        [Description("The unique bricklet device UID.")]
+        public string Uid { get; set; }
```
Missing `TypeConverter`?

glopesdev commented on Bonsai.Tinkerforge/BrickletCO2V2.cs:
```diff
@@ -6,12 +6,12 @@
     {
-        [Description("The unique bricklet device UID.")]
+        [Description("Device data including address UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,211 @@
+namespace Bonsai.Tinkerforge
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Measures coordinates, altitude, modtion, timestamp, satellite status from a GPS Bricklet 2.0.")]
```
Typo in motion, but do we actually measure motion, or just the GPS coordinates? Maybe we revise this to just say something like "Measures GPS coordinates, altitude, satellite status and timer data?".

Also it might be easier to use if the order of the data in the tuple matches the documentation, or if we create a helper class like DataFrame where each data is named.

Finally, see below for a suggestion on different implementation approach.

RoboDoig replied:
There is definitely a reference to motion data in the documentation, gives us course and speed

RoboDoig replied:
Sorry I completely misread this, you're right it is available but we don't use it, will change

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,211 @@
+    public class BrickletGPSV2 : Combinator<IPConnection, Tuple<BrickletGPSV2.StatusData, BrickletGPSV2.CoordinateData, BrickletGPSV2.AltitudeData, BrickletGPSV2.DateTimeData>>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,211 @@
+        public long Period { get; set; } = 1000;
+
+        [Description("If SBAS is enabled, the position accuracy increases (if SBAS satellites are in view), but the update rate is limited to 5Hz. With SBAS disabled the update rate is increased to 10Hz.")]
+        public SBASConfig SBAS { get; set; } = SBASConfig.Disabled;
```
It might be helpful if we spell-out what the SBAS acronym stands for at the beginning of the documentation string. Alternatively, we can include it in the extended documentation later, in which case it might be fine to leave it like this for now.

glopesdev commented on Bonsai.Tinkerforge/BrickletGPSV2.cs:
```diff
@@ -0,0 +1,211 @@
+                    {
+                        try { device.SetDateTimeCallbackPeriod(0); }
+                        catch (NotConnectedException) { }
+                        device.DateTimeCallback -= handler;
+                    });
+                });
+            });
+
+            // Coordinate stream
+            IObservable<CoordinateData> coordinateStream = source.SelectStream(connection =>
+            {
+                var device = new global::Tinkerforge.BrickletGPSV2(Uid, connection);
+                connection.Connected += (sender, e) =>
+                {
+                    device.SetStatusLEDConfig((byte)StatusLed);
+                    device.SetSBASConfig((byte)SBAS);
+                    device.SetCoordinatesCallbackPeriod(Period);
+                };
+
+                return Observable.Create<CoordinateData>(observer =>
+                {
+                    observer.OnNext(new CoordinateData()); // Initialize an empty data struct in case one of the data providers doesn't return anything to the stream combination
+
+                    global::Tinkerforge.BrickletGPSV2.CoordinatesEventHandler handler = (sender, latitude, ns, longitude, ew) =>
+                    {
+                        observer.OnNext(new CoordinateData(latitude, longitude, ns, ew));
+                    };
+
+                    device.CoordinatesCallback += handler;
+                    return Disposable.Create(() =>
+                    {
+                        try { device.SetCoordinatesCallbackPeriod(0); }
+                        catch (NotConnectedException) { }
+                        device.CoordinatesCallback -= handler;
+                    });
+                });
+            });
+
+            return statusStream.CombineLatest(coordinateStream, altitudeStream, dateTimeStream, (s1, s2, s3, s4) => Tuple.Create(s1, s2, s3, s4));
+        }
```
Assuming that reusing the device UID works for the same IPConnection, a totally different, but maybe more flexible approach could be to simply have one operator for each GPSV2 data stream, i.e. literally one node for Altitude, one node for Coordinates, one node for Status, etc.

That would give us maximum flexibility to downstream combine everything as we want, or to just use each stream independently, etc.

glopesdev replied:
If we go this route, it might be useful to create a helper abstract class to contain common functionality such as status code, etc and just override to pick which stream you want to return.

RoboDoig replied:
So far it doesn't appear that the same UID can be reused easily, but I will investigate further

RoboDoig replied:
From trying this out in a non-reactive setting, it appears that the original DeviceReplaced error was a threading problem since we can create two bricklet instances with the same Uid without error, so we can fix this with a lock somewhere in the observable generation. The problem is that each new instance of the bricklet resets the callback definitions. For example:

```
// Generate ip connection
IPConnection ipcon = new IPConnection();
ipcon.Connect("localhost", 4223);

// Generate device
BrickletCO2V2 device = new BrickletCO2V2("Lhu", ipcon);
device.CO2ConcentrationCallback += ConcentrationCB;
device.SetCO2ConcentrationCallbackConfiguration(1000, false, 'x', 0, 0);

// Generate another device
BrickletCO2V2 device2 = new BrickletCO2V2("Lhu", ipcon);
device2.HumidityCallback += HumidityCB;
device2.SetHumidityCallbackConfiguration(1000, false, 'x', 0, 0);

Console.ReadLine();

static void ConcentrationCB(BrickletCO2V2 sender, int concentration)
{
    Console.WriteLine(sender.ToString() + " conce: " + concentration);
}

static void HumidityCB(BrickletCO2V2 sender, int humidity)
{
    Console.WriteLine(sender.ToString() + " humidity: " + humidity);
}
```

This will run fine, but only the 2nd humidity callback will be present in the output. This is only if we use the same IPConnection instance for both Bricklet instances. If we add a second IPConnection instance (with the same host and port) and connect the Bricklets to their own instance of IPConnection both callbacks are present in the output. So possible approaches are:

1. Always create a new IPConnection instance with the same host and port for each Bricklet node
2. Create a new IPConnection instance only when we are attempting to address a bricklet with repeated UID. This requires knowing what bricklets are already on the connection which is not exposed from IPConnection apart from using the EnumerateCallback (which we do in the TypeConverter already)
3. Leave things as they are, 1 IPConnection per unique bricklet and combine streams where multiple data frames are produced.

glopesdev commented on Bonsai.Tinkerforge/BrickletHumidityV2.cs:
```diff
@@ -0,0 +1,80 @@
+    public class BrickletHumidityV2 : Combinator<IPConnection, int>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletHumidityV2.cs:
```diff
@@ -0,0 +1,80 @@
+        public long Period { get; set; } = 1000;
+
+        [Description("Specifies the heater configuration. The heater can be used to dry the sensor in extremely wet conditions")]
+
```
Missing period at end of sentence and extra empty line.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+    public class BrickletIndustrialAnalogOutV2
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+        public string Uid { get; set; }
+
+        [Description("Specifies the output voltage range")]
+        public VoltageRangeConfig VoltageRange { get; set; } = VoltageRangeConfig.Range0To10V;
```
Missing period at end of sentence.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialAnalogOutV2.cs:
```diff
@@ -0,0 +1,90 @@
+        public VoltageRangeConfig VoltageRange { get; set; } = VoltageRangeConfig.Range0To10V;
+
+        [Description("Specifies the output current range")]
+        public CurrentRangeConfig CurrentRange { get; set; } = CurrentRangeConfig.Range0To24mA;
```
Missing period at end of sentence.

glopesdev commented on Bonsai.Tinkerforge/BrickletIndustrialPTC.cs:
```diff
@@ -0,0 +1,80 @@
+    public class BrickletIndustrialPTC : Combinator<IPConnection, int>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletParticulateMatter.cs:
```diff
@@ -0,0 +1,74 @@
+    public class BrickletParticulateMatter : Combinator<IPConnection, BrickletParticulateMatter.DataFrame>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletSoundPressureLevel.cs:
```diff
@@ -0,0 +1,85 @@
+    public class BrickletSoundPressureLevel : Combinator<IPConnection, int>
+    {
+        [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/BrickletThermocoupleV2.cs:
```diff
@@ -11,6 +11,7 @@ namespace Bonsai.Tinkerforge
     public class BrickletThermocoupleV2
     {
         [Description("The unique bricklet device UID.")]
+        [TypeConverter(typeof(UidConverter))]
```
Same as before.

glopesdev commented on Bonsai.Tinkerforge/DeviceData.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <summary>
+    /// Data representation of a connected TinkerForge module
+    /// </summary>
+    public class DeviceData
```
Since we don't seem to be serializing this anymore, can we make this class internal?

glopesdev replied:
Also we could move this class definition inside the `UidConverter` class and rename this file as either `TinkerforgeDeviceLookup.cs` or maybe just move both of these types inside `UidConverter`, as they seem to be implementation details which are only used there anyway?

glopesdev commented on Bonsai.Tinkerforge/DeviceData.cs:
```diff
@@ -0,0 +1,60 @@
+        public int DeviceIdentifier; // Number corresponding to device name
+
+        // Parameterless constructor required for serialization
+        public DeviceData()
```
Since we don't seem to be serializing this anymore, do we still need the parameterless constructor?

glopesdev commented on Bonsai.Tinkerforge/GPSTimerTest.cs:
```diff
@@ -0,0 +1,121 @@
+{
+    [DefaultProperty(nameof(Uid))]
+    [Description("Debug Timer from a GPS Bricklet 2.0.")]
+    public class GPSTimerTest : Combinator<IPConnection, DateTimeData>
```
If we are able to split the data streams of GPSV2 into separate operators as suggested above, then this operator might not be necessary (or could be renamed to something like GPSV2DateTime or something like that).

glopesdev commented on Bonsai.Tinkerforge/UidConverter.cs:
```diff
@@ -0,0 +1,111 @@
+    internal class UidConverter : StringConverter
+    {
+        // Cache for discovered Brick / Bricklet devices
+        public Dictionary<string, DeviceData> devices;
```
Can we move the `DeviceData` class definition inside the `UidConverter` class? It doesn't seem to be used anywhere else so it would make it more explicit that it is an implementation detail.

Also, do we need this member to be `public`?

glopesdev reviewed (changes requested):
Based on your experiments I suggested a different approach to the problem of the GPSV2 streams. Otherwise just small refactoring.

RoboDoig committed:
Fix typos, aesthetic changes, more descriptive comments

RoboDoig committed:
Reorder GPS structs to match documentation

RoboDoig committed:
GPS main description fix

RoboDoig committed:
Test date-time data stream

RoboDoig committed:
DeviceData does not neet to be serialized anymore

RoboDoig committed:
Refactor DeviceData into UidConverter
Device data and lookup only ever used inside the UidConverter

RoboDoig committed:
Cache dictionary need not be public

RoboDoig committed:
Change bricklet naming convention
-->remove global namespace references

RoboDoig committed:
New structure for GPS module
Device initialized in its own nodes, separate streams compartmentalized out

RoboDoig committed:
Altitude and DateTime streams

RoboDoig committed:
Override ToString for string matching in UID search
Alternatively, could have all bricklets inherit from a common parent class, or implement some ISearchable interface with a name lookup

RoboDoig committed:
Analog out V3 bricklet

RoboDoig committed:
Try with separate device creator

RoboDoig committed:
gps callback periods must be set in main node
Otherwise connection is not established before trying to set the callback.

RoboDoig committed:
Use CombineLatest for AnalogOut
Combination of IPConnection observable and signal int observabke

RoboDoig committed:
Merge branch 'analog-out-test' into add-tf-modules

RoboDoig committed:
fix typo

RoboDoig committed:
Industrial analog out using CombineLatest pattern
To merge IPConnection and input signal streams

RoboDoig committed:
Analog out combinelatest --> selectmany alternative

glopesdev committed:
Rewrite to use input signal as optional source

RoboDoig committed:
Industrial analog out same pattern as analog out v3

glopesdev committed:
Avoid lookup delay on bricklet node selection

glopesdev committed:
Merge branch 'add-tf-modules' of https://github.com/emotional-cities/tinkerforge into add-tf-modules

glopesdev committed:
Fix property declaration

RoboDoig committed:
Update initial voltage description for analog outs
